src/arch/riscv/irqc/aia/vaplic.c

```diff
@@ -0,0 +1,1057 @@
+/**
+ * @brief Read from a given interrupt sourcecfg register
+ * 
+ * @param vcpu 
+ * @param intp_id Interrupt id to read
+ * @return uint32_t 
+ */
+static uint32_t vaplic_get_sourcecfg(struct vcpu *vcpu, irqid_t intp_id){
+    uint32_t real_int_id = intp_id - 1;
+    uint32_t ret = 0;
+
+    if(intp_id == 0)
+        return ret;
+    struct virqc * virqc = &vcpu->vm->arch.virqc;
+    if (real_int_id < APLIC_MAX_INTERRUPTS) ret = virqc->srccfg[real_int_id];
+    return ret;
+}
+
+/**
+ * @brief Write to sourcecfg register of a given interrupt
+ * 
+ * @param vcpu 
+ * @param intp_id interrupt id to write to
+ * @param new_val value to write to sourcecfg
+ */
+static void vaplic_set_sourcecfg(struct vcpu *vcpu, irqid_t intp_id, uint32_t new_val){
+    struct virqc *virqc = &vcpu->vm->arch.virqc;
+    spin_lock(&virqc->lock);
+    /** If intp is valid and new source config is different from prev. one.*/
+    if (intp_id > 0 && intp_id < APLIC_MAX_INTERRUPTS && 
+        vaplic_get_sourcecfg(vcpu, intp_id) != new_val) {
+        /** If intp is being delegated make whole reg 0.
+         *  This happens because this domain is always a leaf. */        
+        new_val &= (new_val & APLIC_SRCCFG_D) ? 0 : APLIC_SRCCFG_SM;
+
+        /** If SM is reserved make intp inactive */
+        if(new_val == 2 || new_val == 3)
+            new_val = APLIC_SOURCECFG_SM_INACTIVE;
+        
+        /** Only edge sense can be virtualized for know */
+        if(new_val  == APLIC_SOURCECFG_SM_LEVEL_HIGH){
+            new_val = APLIC_SOURCECFG_SM_EDGE_RISE;
+        } else if (new_val  == APLIC_SOURCECFG_SM_LEVEL_LOW){
+            new_val = APLIC_SOURCECFG_SM_EDGE_FALL;
+        }
+
+        /** Is this intp a phys. intp? */
+        if(vaplic_get_hw(vcpu, intp_id)){
+            /** Update in phys. aplic */
+            aplic_set_sourcecfg(intp_id, new_val);
+            /** If phys aplic was succe. written, then update virtual*/
+            if(aplic_get_sourcecfg(intp_id) == new_val){
+                virqc->srccfg[intp_id-1] = new_val;
+            }
+        } else {
+            /** If intp is not phys. update virtual only */
+            virqc->srccfg[intp_id-1] = new_val;
+        }
+        vaplic_update_hart_line(vcpu);
+    }
+    spin_unlock(&virqc->lock);
+}
+
+/**
+ * @brief Get the pending bits for interrupts [32*reg:(32*reg)+31]
+ * 
+ * @param vcpu virtual cpu
+ * @param reg regiter index
+ * @return uint32_t value with pending bit mapped per bit
+ */
+static uint32_t vaplic_get_setip(struct vcpu *vcpu, uint8_t reg){
+    uint32_t ret = 0;
+    struct virqc * virqc = &vcpu->vm->arch.virqc;
+    if (reg < APLIC_NUM_SETIx_REGS) ret = virqc->setip[reg];
+    return ret;
+}
+
+/**
+ * @brief Set the pending bits for interrupts [32*reg:(32*reg)+31]
+ * 
+ * @param vcpu virtual cpu
+ * @param reg regiter index
+ * @param new_val value with pending bit mapped per bit
+ */
+static void vaplic_set_setip(struct vcpu *vcpu, uint8_t reg, uint32_t new_val){
```

Comment:
This function has changed a lot. I am putting bellow the current code:

```
static void vaplic_set_setip(struct vcpu *vcpu, uint8_t reg, uint32_t new_val){
    struct vaplic *vaplic = &vcpu->vm->arch.vaplic;

    spin_lock(&vaplic->lock);
    if (reg == 0) new_val &= MASK_INTP_ZERO;
    if (reg < APLIC_NUM_SETIx_REGS) {
        for(size_t i = (reg*APLIC_NUM_INTP_PER_REG); 
            i < (reg*APLIC_NUM_INTP_PER_REG) + APLIC_NUM_INTP_PER_REG; i++){
            if (!!bit32_get(new_val, i%32)){
                if(vaplic_set_pend(vcpu, i)){
                    vaplic_update_hart_line(vcpu, GET_HART_INDEX(vcpu, i));
                }
            }
        }
    }
    spin_unlock(&vaplic->lock);
}
```

My question here is about a design choice. We can make this function cleaner. However, we will not be able to update only the harts that have suffered changes. The alternative code is:

```
static void vaplic_set_setip(struct vcpu *vcpu, uint8_t reg, uint32_t new_val){
    struct vaplic *vaplic = &vcpu->vm->arch.vaplic;

    spin_lock(&vaplic->lock);
    if (reg == 0) new_val &= MASK_INTP_ZERO;
    if (reg < APLIC_NUM_SETIx_REGS) {
        vaplic->ip[reg] = new_val & vaplic->active[reg];
        vaplic_update_hart_line(vcpu, UPDATE_ALL_HARTS);
    }
    spin_unlock(&vaplic->lock);
}
```

As you can see, since we are updating the whole register at once, we don't know exactly which interrupts have changed their pending state.

What do you think? Do we do the simpler code, updating all the hart lines (even though there is a possibility that only one hart has changed), or do we keep the previous code?

Reply:
Please take a look at commit 9b4d74b22b216ed9e2896cb4b08bf72763973454. What do you think?